src/query/typed_value.cpp

```diff
@@ -362,6 +362,33 @@ bool TypedValue::IsGraph() const { return type_ == Type::Graph; }
 
 #undef DEFINE_VALUE_AND_TYPE_GETTERS
 
+bool TypedValue::ContainsDeleted() const {
+  switch (type_) {
+    case Type::List:
+      for (const auto &elem : list_v) {
+        if (elem.ContainsDeleted()) return true;
+      }
+      return false;
+    case Type::Map:
+      for (const auto &[_, map_value] : map_v) {
+        if (map_value.ContainsDeleted()) return true;
+      }
+      return false;
+    case Type::Vertex:
+      return vertex_v.impl_.vertex_->deleted;
+    case Type::Edge:
+      return edge_v.impl_.edge_.ptr->deleted;
```

Comment:
Should we here handle a case when Edge doesn't have properties? In this case ptr should be invalid.

Reply:
Added. In solving this, my original idea was to add a `deleted` property to `EdgeRef`, but it wasn’t possible to update that property in storage.cpp because an `EdgeRef` object isn’t unique to its edge.

Because of this, there’s no way to check whether an `EdgeRef` refers to an existing edge if `ptr` is invalid (short of major changes to storage).